src/main/java/org/truffleruby/language/globals/ReadMatchReferenceNodes.java

```diff
@@ -0,0 +1,319 @@
+        }
+
+        @TruffleBoundary
+        private void setNamedForNonNil(MaterializedFrame frame, Regex regex, DynamicObject string, DynamicObject matchData) {
+            DynamicObject nil = nil();
+            Region region = Layouts.MATCH_DATA.getRegion(matchData);
+            for (Iterator<NameEntry> i = regex.namedBackrefIterator(); i.hasNext();) {
+                final NameEntry e = i.next();
+                final String name = new String(e.name, e.nameP, e.nameEnd - e.nameP, StandardCharsets.UTF_8).intern();
+                int nth = regex.nameToBackrefNumber(e.name, e.nameP, e.nameEnd, region);
+
+                final Object value;
+
+                // Copied from jruby/RubyRegexp - see copyright notice there
+
+                if (nth >= region.numRegs || (nth < 0 && (nth += region.numRegs) <= 0)) {
+                    value = nil;
+                } else {
+                    final int start = region.beg[nth];
+                    final int end = region.end[nth];
+                    if (start != -1) {
+                        value = createSubstring(string, start, end - start);
+                    } else {
+                        value = nil;
+                    }
+                }
+
+                setLocalVariable(frame, name, value);
+            }
+        }
+
+        @TruffleBoundary
+        private void setNamedForNil(MaterializedFrame frame) {
+            DynamicObject nil = nil();
+            for (Iterator<NameEntry> i = regex.namedBackrefIterator(); i.hasNext();) {
+                final NameEntry e = i.next();
+                final String name = new String(e.name, e.nameP, e.nameEnd - e.nameP, StandardCharsets.UTF_8).intern();
+                setLocalVariable(frame, name, nil);
+            }
+        }
+
+        private static void setLocalVariable(MaterializedFrame frame, String name, Object value) {
+            assert value != null;
+
+            while (frame != null) {
+                final FrameSlot slot = frame.getFrameDescriptor().findFrameSlot(name);
+                if (slot != null) {
+                    frame.setObject(slot, value);
+                    break;
+                }
+
+                frame = RubyArguments.getDeclarationFrame(frame);
+            }
+        }
+
+        private DynamicObject createSubstring(DynamicObject source, int start, int length) {
+            assert RubyGuards.isRubyString(source);
+
+            final Rope sourceRope = StringOperations.rope(source);
+            final Rope substringRope = makeSubstringNode.executeMake(sourceRope, start, length);
+
+            final DynamicObject ret = Layouts.CLASS.getInstanceFactory(Layouts.BASIC_OBJECT.getLogicalClass(source)).newInstance(Layouts.STRING.build(false, false, substringRope));
+
+            return ret;
+        }
+    }
+
+    public static RubyNode createNamePopulateingNode(Regex regex, RubyNode matchResultNode, GetFromThreadAndFrameLocalStorageNode readMatchNode) {
```

Comment:
Type: `Populating`, but I don't think this needs a factory method at all, the node name is good enough.

Reply:
Yeah, I had put a factory method in when I thought I'd do more optimal name setting as part of this change set.